### PR TITLE
Update to use newer API

### PR DESCRIPTION
**Pull request by chrisbdaemon**
The version of the API being used here isn't very safe.  Its mitm-able so heres some changes that should bring it up to the newer authentication api that uses hmac with a pre-shared key to verify the message integrity.

### CODE REVIEW

File: README.rdoc [hunk 1]
```diff
@@ -22,7 +22,9 @@ A library to verify, decode, decrypt and parse Yubikey[http://www.yubico.com/hom
 === OTP Verification
 
   begin
-    otp = Yubikey::OTP::Verify.new('dteffujehknhfjbrjnlnldnhcujvddbikngjrtgh')
+    otp = Yubikey::OTP::Verify.new(:id => 1234,
+                                   :key => 'NiSwGZBQ0gTbwXbRGWAf4kM5xXg=',
```
Comment (titanous) [inline README.rdoc:26]: Please rename these parameters to `:api_id` and `:api_key`.

File: lib/yubikey/otp_verify.rb [hunk 2]
```diff
@@ -1,16 +1,30 @@
...
-    
     # The raw status from the Yubico server
     attr_reader :status
-    
-    def initialize(otp)
-      verify("id=#{API_ID}&otp=#{otp}")
+
+    @id  = ''
+    @key = ''
+    @http
```
Comment (titanous) [inline lib/yubikey/otp_verify.rb]: What do these do?

File: lib/yubikey/otp_verify.rb [hunk 3]
```diff
@@ -33,12 +48,55 @@ def verify(query)
...
+      if ! verify_response(result)
+        @status = 'BAD_RESPONSE'
+        return
+      end
+    end
+
+    def generate_nonce(length)
+      characters = ("a".."z").to_a + ("A".."Z").to_a + ("0".."9").to_a
+      nonce = ''
+      1.upto(length) { |index| nonce << characters[rand(characters.size-1)] }
```
Comment (titanous) [inline lib/yubikey/otp_verify.rb]: Just use `SecureRandom.hex`.

File: lib/yubikey/otp_verify.rb [hunk 4]
```diff
@@ -1,16 +1,30 @@
...
+      raise(ArgumentError, "Must supply API Key") if args[:key].nil?
+      raise(ArgumentError, "Must supply OTP") if args[:otp].nil?
+
+      @key = args[:key]
+      @id = args[:id]
+      
+      @url = args[:url] || API_URL
+      @nonce = args[:nonce] || generate_nonce(32)
+
+      verify("id=#{@id}&otp=#{args[:otp]}")
```
Comment (titanous) [inline lib/yubikey/otp_verify.rb]: Pass the parameters to verify as a hash, and build the query there.

File: lib/yubikey/otp_verify.rb [hunk 5]
```diff
@@ -33,12 +48,55 @@ def verify(query)
...
+      if @nonce != returned_nonce
+        return false
+      end
+
+      generated_signature = self.class.generate_hmac(result, @key)
+
+      return signature == generated_signature
+    end
+
+    def self.generate_hmac(response, key)
```
Comment (titanous) [inline lib/yubikey/otp_verify.rb]: This should just be private instance method.

**Comment by titanous** [comment]
Thanks!

I haven't touched this in a few years. I'll merge it in and release if you can make a few style changes that I pointed out.

**Comment by dennisreimann** [comment]
Nice, I'm also looking forward to see that integrated :)

**Comment by chrisbdaemon** [comment]
Let me know if this is alright.

**Comment by wndxlori** [comment]
Is there anything left to do here? Would be nice to see this merged.

**Comment by ukd1** [comment]
@wndxlori would you be able to update this so it can be merged / specs pass?

**Comment by wndxlori** [comment]
I can certainly try, since I'm at the point of actually needing to integrate Yubikey in a client project _right now_ :-)

**Comment by ukd1** [comment]
@wndxlori great! I'm hoping to get #10 fixed this weekend. Let me know if I can help.

**Comment by ukd1** [comment]
@wndxlori I didn't notice that @chrisbdaemon wrote the original code....oops. Feel free to if you want, if not @chrisbdaemon?

**Comment by maartenvg** [comment]
@ukd1 #10 is fixed. Default encoding for Ruby 2.0.0 is now UTF-8 (Yay!) but that means the strings in the specs will be parsed incorrectly.

**Comment by maartenvg** [comment]
I updated the code from @chrisbdaemon for this pull-request, cleaned it a bit up and merged it in master.

[edit] Mentioned the wrong name, sorry dennis [/edit]

**Pull request merged**